File: packages/jetbrains-plugin/src/main/kotlin/com/actionlens/logs/ErrorDetector.kt
package com.actionlens.logs

enum class LineSeverity { ERROR, WARNING, INFO }

private data class Pattern(val id: String, val severity: LineSeverity, val regex: Regex)

private val TIMESTAMP_RE = Regex("""^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z\s""")

private val PATTERNS = listOf(
    // Warnings first.
    Pattern("warning-prefix", LineSeverity.WARNING, Regex("""^(?:\s*\[?)?warning\b""", RegexOption.IGNORE_CASE)),
    Pattern("warn-prefix", LineSeverity.WARNING, Regex("""^(?:\s*\[?)?warn\b""", RegexOption.IGNORE_CASE)),
    Pattern("deprecated", LineSeverity.WARNING, Regex("""\bdeprecat(?:ed|ion)\b""", RegexOption.IGNORE_CASE)),

    // GitHub Actions infra.
    Pattern("process-exit", LineSeverity.ERROR, Regex("""process completed with exit code\s+[1-9]""", RegexOption.IGNORE_CASE)),
    Pattern("github-error-prefix", LineSeverity.ERROR, Regex("""^##\[error\]""", RegexOption.IGNORE_CASE)),

    // JS / Node tooling.
    Pattern("npm-err", LineSeverity.ERROR, Regex("""^npm\s+ERR!""")),
    Pattern("pnpm-err", LineSeverity.ERROR, Regex("""^\s*(?:ELIFECYCLE|ERR_PNPM_[A-Z0-9_]+)\b""")),
    Pattern("yarn-error", LineSeverity.ERROR, Regex("""^(?:yarn\s+)?error\s""", RegexOption.IGNORE_CASE)),

    // PHP / Laravel.
    Pattern("phpunit-fail", LineSeverity.ERROR, Regex("""\bFAILURES!\b|^FAILED\b""")),
    Pattern("pest-fail", LineSeverity.ERROR, Regex("""^\s*(?:FAIL|FAILED)\s+Tests\\""")),
    Pattern("laravel-exception", LineSeverity.ERROR, Regex("""\b(?:Illuminate\\|Symfony\\Component\\).+Exception\b""")),
    Pattern("composer-error", LineSeverity.ERROR, Regex("""^\s*\[?(?:RuntimeException|ErrorException)\]?\b""")),
    Pattern("sqlstate", LineSeverity.ERROR, Regex("""\bSQLSTATE\[[A-Z0-9]+]""")),

    // TS / ESLint.
    Pattern("ts-error", LineSeverity.ERROR, Regex(""":\s*error TS\d+:""")),
    Pattern("eslint-error", LineSeverity.ERROR, Regex("""^\s*\d+:\d+\s+error\s""", RegexOption.IGNORE_CASE)),

    // Docker.
    Pattern("docker-build-failed", LineSeverity.ERROR, Regex("""^(?:ERROR:\s+)?failed to (?:build|solve)""", RegexOption.IGNORE_CASE)),
    Pattern("docker-no-such-image", LineSeverity.ERROR, Regex("""^(?:Error response from daemon|docker:)\s+""", RegexOption.IGNORE_CASE)),

    // Shell.
    Pattern("permission-denied", LineSeverity.ERROR, Regex("""\bpermission denied\b""", RegexOption.IGNORE_CASE)),
    Pattern("command-not-found", LineSeverity.ERROR, Regex("""\b(?:command not found|No such file or directory)\b""", RegexOption.IGNORE_CASE)),

    // Generic.
    Pattern("traceback", LineSeverity.ERROR, Regex("""^Traceback \(most recent call last\):""")),
    Pattern("exception", LineSeverity.ERROR, Regex("""^\s*[\w.]*Exception(?:\s*[:\[]|\s+in\s)""")),
    Pattern("fatal", LineSeverity.ERROR, Regex("""^\s*fatal[:\s]""", RegexOption.IGNORE_CASE)),
    Pattern("error-prefix", LineSeverity.ERROR, Regex("""^(?:\s*\[?)?error[:\s]""", RegexOption.IGNORE_CASE)),
    Pattern("failed-prefix", LineSeverity.ERROR, Regex("""^(?:\s*\[?)?failed[:\s]""", RegexOption.IGNORE_CASE)),
)

object ErrorDetector {
    fun stripTimestamp(line: String): String = TIMESTAMP_RE.replace(line, "")

    fun classify(line: String): Pair<LineSeverity, String?> {
        val cleaned = stripTimestamp(line)
        for (p in PATTERNS) {
            if (p.regex.containsMatchIn(cleaned)) return p.severity to p.id
        }
        return LineSeverity.INFO to null
    }
}

File: packages/core/src/logs/errorDetector.ts
import type { LineSeverity } from "./types";

interface Pattern {
  id: string;
  severity: LineSeverity;
  regex: RegExp;
}

/**
 * Ordered list of patterns. First match wins, so put more-specific patterns
 * before generic ones (e.g. "warning" before "error").
 *
 * We deliberately match on word boundaries / line starts where possible to
 * keep false-positive rate low — many tools print decorative "Error:" headers
 * inside successful output.
 */
const PATTERNS: Pattern[] = [
  // Warnings first — many tools print "warning: this is not an error".
  { id: "warning-prefix", severity: "warning", regex: /^(?:\s*\[?)?warning\b/i },
  { id: "warn-prefix", severity: "warning", regex: /^(?:\s*\[?)?warn\b/i },
  { id: "deprecated", severity: "warning", regex: /\bdeprecat(?:ed|ion)\b/i },

  // Hard failures that GitHub itself emits.
  { id: "process-exit", severity: "error", regex: /process completed with exit code\s+[1-9]/i },
  { id: "github-error-prefix", severity: "error", regex: /^##\[error\]/i },

  // JS / Node tooling.
  { id: "npm-err", severity: "error", regex: /^npm\s+ERR!/ },
  { id: "pnpm-err", severity: "error", regex: /^\s*(?:ELIFECYCLE|ERR_PNPM_[A-Z0-9_]+)\b/ },
  { id: "yarn-error", severity: "error", regex: /^(?:yarn\s+)?error\s/i },

  // PHP / Laravel / PHPUnit / Pest.
  { id: "phpunit-fail", severity: "error", regex: /\bFAILURES!\b|^FAILED\b/ },
  { id: "pest-fail", severity: "error", regex: /^\s*(?:FAIL|FAILED)\s+Tests\\/ },
  { id: "laravel-exception", severity: "error", regex: /\b(?:Illuminate\\|Symfony\\Component\\).+Exception\b/ },
  { id: "composer-error", severity: "error", regex: /^\s*\[?(?:RuntimeException|ErrorException)\]?\b/ },
  { id: "sqlstate", severity: "error", regex: /\bSQLSTATE\[[A-Z0-9]+\]/ },

  // TypeScript / ESLint.
  { id: "ts-error", severity: "error", regex: /:\s*error TS\d+:/ },
  { id: "eslint-error", severity: "error", regex: /^\s*\d+:\d+\s+error\s/i },

  // Docker.
  { id: "docker-build-failed", severity: "error", regex: /^(?:ERROR:\s+)?failed to (?:build|solve)/i },
  { id: "docker-no-such-image", severity: "error", regex: /^(?:Error response from daemon|docker:)\s+/i },

  // Shell-level failures.
  { id: "permission-denied", severity: "error", regex: /\bpermission denied\b/i },
  { id: "command-not-found", severity: "error", regex: /\b(?:command not found|No such file or directory)\b/i },

  // Generic last resorts.
  { id: "traceback", severity: "error", regex: /^Traceback \(most recent call last\):/ },
  { id: "exception", severity: "error", regex: /^\s*[\w.]*Exception(?:\s*[:[]|\s+in\s)/ },
  { id: "fatal", severity: "error", regex: /^\s*fatal[:\s]/i },
  { id: "error-prefix", severity: "error", regex: /^(?:\s*\[?)?error[:\s]/i },
  { id: "failed-prefix", severity: "error", regex: /^(?:\s*\[?)?failed[:\s]/i },
];

const TIMESTAMP_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z\s/;

export function stripTimestamp(line: string): string {
  return line.replace(TIMESTAMP_RE, "");
}

export function classify(line: string): { severity: LineSeverity; matchedPattern?: string } {
  const cleaned = stripTimestamp(line);
  for (const pattern of PATTERNS) {
    if (pattern.regex.test(cleaned)) {
      return { severity: pattern.severity, matchedPattern: pattern.id };
    }
  }
  return { severity: "info" };
}
